src/types/function/mod.rs

```diff
@@ -0,0 +1,217 @@
+    pub(crate) this: Option<Handle<'a, JsValue>>,
+    pub(crate) args: private::ArgsVec<'a>,
+}
+
+impl<'a> CallOptions<'a> {
+    /// Set the value of `this` for the function call.
+    pub fn this<V: Value>(&mut self, this: Handle<'a, V>) -> &mut Self {
+        self.this = Some(this.upcast());
+        self
+    }
+
+    /// Add an argument to the arguments list.
+    pub fn arg<V: Value>(&mut self, arg: Handle<'a, V>) -> &mut Self {
+        self.args.push(arg.upcast());
+        self
+    }
+
+    /// Replaces the arguments list with the given arguments.
+    pub fn args<A: Arguments<'a>>(&mut self, args: A) -> &mut Self {
+        self.args = args.into_args_vec();
+        self
+    }
+
+    /// Make the function call. If the function returns without throwing, the result value
+    /// is downcast to the type `V`, throwing a `TypeError` if the downcast fails.
+    pub fn apply<'b: 'a, V: Value, C: Context<'b>>(&self, cx: &mut C) -> JsResult<'b, V> {
+        let this = self.this.unwrap_or_else(|| cx.null().upcast());
+        let v: Handle<JsValue> = self.callee.call(cx, this, &self.args)?;
+        v.downcast_or_throw(cx)
+    }
+
+    /// Make the function call for side effect, discarding the result value. This method is
+    /// preferable to [`apply()`](CallOptions::apply) when the result value isn't needed,
+    /// since it doesn't require specifying a result type.
+    pub fn exec<'b: 'a, C: Context<'b>>(&self, cx: &mut C) -> NeonResult<()> {
+        let this = self.this.unwrap_or_else(|| cx.null().upcast());
+        self.callee.call(cx, this, &self.args)?;
+        Ok(())
+    }
+}
+
+/// A builder for making a JavaScript constructor call like `new Array(16)`.
+///
+/// The builder methods make it convenient to assemble the call from parts:
+/// ```
+/// # use neon::prelude::*;
+/// # fn foo(mut cx: FunctionContext) -> JsResult<JsObject> {
+/// # let global = cx.global();
+/// # let url = global.get(&mut cx, "URL")?;
+/// # let url: Handle<JsFunction> = url.downcast_or_throw(&mut cx)?;
+/// let obj = url
+///     .construct_with(&cx)
+///     .arg(cx.string("https://neon-bindings.com"))
+///     .apply(&mut cx)?;
+/// # Ok(obj)
+/// # }
+/// ```
+#[derive(Clone)]
+pub struct ConstructOptions<'a> {
+    pub(crate) callee: Handle<'a, JsFunction>,
+    pub(crate) args: private::ArgsVec<'a>,
+}
+
+impl<'a> ConstructOptions<'a> {
+    /// Add an argument to the arguments list.
+    pub fn arg<V: Value>(&mut self, arg: Handle<'a, V>) -> &mut Self {
+        self.args.push(arg.upcast());
+        self
+    }
+
+    /// Replaces the arguments list with the given arguments.
+    pub fn args<A: Arguments<'a>>(&mut self, args: A) -> &mut Self {
+        self.args = args.into_args_vec();
+        self
+    }
+
+    /// Make the constructor call. If the function returns without throwing, returns
+    /// the resulting object.
+    pub fn apply<'b: 'a, C: Context<'b>>(&self, cx: &mut C) -> JsResult<'b, JsObject> {
```

Comment:
How come we don't downcast on this one?

```rust
let global = cx.global();
let array_ctor: Handle<JsFunction> = global.get(&mut cx, "Array");
let arr: Handle<JsArray> = array_ctor.construct_with(&cx).arg(cx.number(42)).apply(&mut cx);
```

A more realistic example would probably be calling a method that returns a `JsBox`.

Reply:
The only downside is that it's usually going to be `JsObject` but you'll sometimes have to manually type-annotate to assist inference:

```rust
let o: Handle<JsObject> = cx
    .argument::<JsFunction>(0)?
    .construct_with(&cx)
    .arg(cx.number(0.0))
    .apply(&mut cx)?;
```

But I think it's worth it so that this API consistently never requires you to use `downcast_or_throw()`.